File: 2025/day09/go/solution.go
package main

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

type Point struct {
	x, y int
}

type Edge struct {
	y, xMin, xMax int // For horizontal edges
}

type VertEdge struct {
	x, yMin, yMax int // For vertical edges
}

func parseInput(filename string) ([]Point, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}

	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	points := make([]Point, 0, len(lines))

	for _, line := range lines {
		parts := strings.Split(line, ",")
		if len(parts) != 2 {
			continue
		}
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		points = append(points, Point{x, y})
	}

	return points, nil
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func part1(points []Point) int {
	maxArea := 0
	n := len(points)

	// Check all pairs of points as opposite corners
	for i := 0; i < n; i++ {
		x1, y1 := points[i].x, points[i].y
		for j := i + 1; j < n; j++ {
			x2, y2 := points[j].x, points[j].y
			// Rectangle area = width * height (inclusive of both corners)
			width := abs(x2-x1) + 1
			height := abs(y2-y1) + 1
			area := width * height
			if area > maxArea {
				maxArea = area
			}
		}
	}

	return maxArea
}

func part2(points []Point) int {
	n := len(points)

	// Build edges
	horizontalEdges := make([]Edge, 0)
	verticalEdges := make([]VertEdge, 0)

	for i := 0; i < n; i++ {
		x1, y1 := points[i].x, points[i].y
		x2, y2 := points[(i+1)%n].x, points[(i+1)%n].y

		if y1 == y2 { // Horizontal edge
			horizontalEdges = append(horizontalEdges, Edge{y1, min(x1, x2), max(x1, x2)})
		} else { // Vertical edge
			verticalEdges = append(verticalEdges, VertEdge{x1, min(y1, y2), max(y1, y2)})
		}
	}

	// Map edges by coordinate for efficient lookup
	vertByX := make(map[int][]VertEdge)
	for _, edge := range verticalEdges {
		vertByX[edge.x] = append(vertByX[edge.x], edge)
	}

	horizByY := make(map[int][]Edge)
	for _, edge := range horizontalEdges {
		horizByY[edge.y] = append(horizByY[edge.y], edge)
	}

	// Check if point is inside polygon using ray casting
	isInsidePolygon := func(x, y float64) bool {
		crossings := 0.0
		// Cast ray to the right
		for vx := range vertByX {
			if float64(vx) <= x {
				continue
			}
			for _, edge := range vertByX[vx] {
				if float64(edge.yMin) < y && y < float64(edge.yMax) {
					// Strict inequality - ray crosses edge
					crossings += 1.0
				} else if y == float64(edge.yMin) || y == float64(edge.yMax) {
					// On corner - count as 0.5 crossing
					crossings += 0.5
				}
			}
		}
		return math.Mod(crossings, 2) == 1
	}

	// Check if rectangle is valid (entirely inside polygon)
	rectangleValid := func(x1, y1, x2, y2 int) bool {
		minX, maxX := min(x1, x2), max(x1, x2)
		minY, maxY := min(y1, y2), max(y1, y2)

		// Check if any vertical edge crosses through the rectangle interior
		for vx, edges := range vertByX {
			if minX < vx && vx < maxX {
				for _, edge := range edges {
					// Check if this edge segment overlaps with rectangle's y range
					if !(edge.yMax <= minY || edge.yMin >= maxY) {
						return false
					}
				}
			}
		}

		// Check if any horizontal edge crosses through the rectangle interior
		for hy, edges := range horizByY {
			if minY < hy && hy < maxY {
				for _, edge := range edges {
					// Check if this edge segment overlaps with rectangle's x range
					if !(edge.xMax <= minX || edge.xMin >= maxX) {
						return false
					}
				}
			}
		}

		// Finally, check that we're inside the polygon (not outside)
		// Check center point
		centerX := float64(minX+maxX) / 2.0
		centerY := float64(minY+maxY) / 2.0
		return isInsidePolygon(centerX, centerY)
	}

	// Find largest valid rectangle with red corners
	maxArea := 0

	for i := 0; i < len(points); i++ {
		x1, y1 := points[i].x, points[i].y
		for j := i + 1; j < len(points); j++ {
			x2, y2 := points[j].x, points[j].y

			if rectangleValid(x1, y1, x2, y2) {
				width := abs(x2-x1) + 1
				height := abs(y2-y1) + 1
				area := width * height
				if area > maxArea {
					maxArea = area
				}
			}
		}
	}

	return maxArea
}

func main() {
	// Get the directory containing this executable
	exePath, err := os.Executable()
	if err != nil {
		fmt.Println("Error getting executable path:", err)
		return
	}
	exeDir := filepath.Dir(exePath)
	inputPath := filepath.Join(exeDir, "..", "input.txt")

	// If running with go run, use current directory
	if strings.Contains(exePath, "go-build") {
		inputPath = "../input.txt"
	}

	points, err := parseInput(inputPath)
	if err != nil {
		fmt.Println("Error reading input:", err)
		return
	}

	fmt.Printf("Part 1: %d\n", part1(points))
	fmt.Printf("Part 2: %d\n", part2(points))
}

File: 2025/day09/rust/solution.rs
use std::collections::HashMap;
use std::fs;

fn main() {
    let input = fs::read_to_string("../input.txt")
        .expect("Failed to read input file");

    let points = parse_input(&input);

    println!("Part 1: {}", part1(&points));
    println!("Part 2: {}", part2(&points));
}

fn parse_input(input: &str) -> Vec<(i32, i32)> {
    input
        .trim()
        .lines()
        .map(|line| {
            let parts: Vec<&str> = line.split(',').collect();
            let x = parts[0].parse::<i32>().unwrap();
            let y = parts[1].parse::<i32>().unwrap();
            (x, y)
        })
        .collect()
}

fn part1(points: &[(i32, i32)]) -> i64 {
    let mut max_area: i64 = 0;
    let n = points.len();

    // Check all pairs of points as opposite corners
    for i in 0..n {
        let (x1, y1) = points[i];
        for j in (i + 1)..n {
            let (x2, y2) = points[j];
            // Rectangle area = width * height (inclusive of both corners)
            let width = ((x2 - x1).abs() + 1) as i64;
            let height = ((y2 - y1).abs() + 1) as i64;
            let area = width * height;
            max_area = max_area.max(area);
        }
    }

    max_area
}

fn part2(points: &[(i32, i32)]) -> i64 {
    let n = points.len();
    let mut horizontal_edges = Vec::new();  // (y, x_min, x_max)
    let mut vertical_edges = Vec::new();    // (x, y_min, y_max)

    // Build edges from consecutive points (forming a polygon)
    for i in 0..n {
        let (x1, y1) = points[i];
        let (x2, y2) = points[(i + 1) % n];

        if y1 == y2 {
            // Horizontal edge
            horizontal_edges.push((y1, x1.min(x2), x1.max(x2)));
        } else {
            // Vertical edge
            vertical_edges.push((x1, y1.min(y2), y1.max(y2)));
        }
    }

    // Build lookup maps for efficient edge queries
    let mut vert_by_x: HashMap<i32, Vec<(i32, i32)>> = HashMap::new();
    for &(x, y_min, y_max) in &vertical_edges {
        vert_by_x.entry(x).or_insert_with(Vec::new).push((y_min, y_max));
    }

    let mut horiz_by_y: HashMap<i32, Vec<(i32, i32)>> = HashMap::new();
    for &(y, x_min, x_max) in &horizontal_edges {
        horiz_by_y.entry(y).or_insert_with(Vec::new).push((x_min, x_max));
    }

    // Helper function to check if a point is inside the polygon using ray casting
    let is_inside_polygon = |x: f64, y: f64| -> bool {
        let mut crossings = 0.0;

        // Get sorted list of vertical edge x-coordinates
        let mut vx_coords: Vec<i32> = vert_by_x.keys().copied().collect();
        vx_coords.sort();

        // Cast ray to the right
        for vx in vx_coords {
            if (vx as f64) <= x {
                continue;
            }
            if let Some(edges) = vert_by_x.get(&vx) {
                for &(y_min, y_max) in edges {
                    if (y_min as f64) < y && y < (y_max as f64) {
                        // Ray crosses edge
                        crossings += 1.0;
                    } else if y == (y_min as f64) || y == (y_max as f64) {
                        // On corner - count as 0.5 crossing
                        crossings += 0.5;
                    }
                }
            }
        }

        crossings % 2.0 == 1.0
    };

    // Helper function to check if rectangle is entirely inside polygon
    let rectangle_valid = |x1: i32, y1: i32, x2: i32, y2: i32| -> bool {
        let min_x = x1.min(x2);
        let max_x = x1.max(x2);
        let min_y = y1.min(y2);
        let max_y = y1.max(y2);

        // Check if any vertical edge crosses through the rectangle interior
        for (&vx, edges) in &vert_by_x {
            if min_x < vx && vx < max_x {
                for &(y_min, y_max) in edges {
                    // Check if this edge segment overlaps with rectangle's y range
                    if !(y_max <= min_y || y_min >= max_y) {
                        return false;
                    }
                }
            }
        }

        // Check if any horizontal edge crosses through the rectangle interior
        for (&hy, edges) in &horiz_by_y {
            if min_y < hy && hy < max_y {
                for &(x_min, x_max) in edges {
                    // Check if this edge segment overlaps with rectangle's x range
                    if !(x_max <= min_x || x_min >= max_x) {
                        return false;
                    }
                }
            }
        }

        // Finally, check that we're inside the polygon (not outside)
        // Check center point
        let center_x = (min_x + max_x) as f64 / 2.0;
        let center_y = (min_y + max_y) as f64 / 2.0;
        is_inside_polygon(center_x, center_y)
    };

    // Find largest valid rectangle with red corners
    let mut max_area: i64 = 0;

    for i in 0..n {
        let (x1, y1) = points[i];
        for j in (i + 1)..n {
            let (x2, y2) = points[j];

            if rectangle_valid(x1, y1, x2, y2) {
                let width = ((x2 - x1).abs() + 1) as i64;
                let height = ((y2 - y1).abs() + 1) as i64;
                let area = width * height;
                max_area = max_area.max(area);
            }
        }
    }

    max_area
}
